Crash in audit log if binary data is passed to authentication code verification

### Affected product(s) and version(s)

develop

### Environment

dev

### Steps to reproduce, current result, expected result

Verify binary data (not JSON) in `/rest/v4/auth/verify`. It seems that verification passed, but the attempt to write entry to the audit log failed. I don't see any limitation in the documentation, that we have to always pass textual content into authentication code verification.

### Log output

```
2025-08-28T13:51:08.739+02:00  INFO 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] w.s.p.a.s.c.a.v.AuthenticationController : action: verifyAuthentication, state: initiated, activationId: 06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0, applicationKey: qjsBj6II/syLPs65BbNRyg==
2025-08-28T13:51:08.739+02:00 DEBUG 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] w.s.p.a.s.c.a.v.AuthenticationController : action: verifyAuthentication, state: initiated, request: ObjectRequest(requestObject=VerifyAuthenticationRequest(activationId=06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0, applicationKey=qjsBj6II/syLPs65BbNRyg==, data=POST&L2hlbGxvL2Jpb2hhY2tlcg==&OBktkhL3t+YAgZlZs81cJQ==&FtXwLcYDqRUA3+rOW4JiegPhhlG+gzRDH8m2kl+UcDNmPxFtkrMnKq8S39WE9b1wMPHGpHsureF8IVVDIB5L, authenticationCodeType=POSSESSION_BIOMETRY, authenticationVersion=4.0, forcedAuthenticationVersion=null, allowedStates=[]))
2025-08-28T13:51:08.781+02:00  INFO 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] org.hibernate.orm.jdbc.batch             : HHH100503: On release of batch it still contained JDBC statements
2025-08-28T13:51:08.782+02:00  WARN 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 22021
2025-08-28T13:51:08.782+02:00 ERROR 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] o.h.engine.jdbc.spi.SqlExceptionHelper   : Batch entry 0 insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (('06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0'),('2'::int8),('xL7oBxcFPIH0IoAFoTm1gQmgmam1UbRZAhFQj9NS6zw='),('3'::int4),('[ ]'),('UE9TVCZMMmhsYkd4dkwySnBiMmhoWTJ0bGNnPT0mT0JrdGtoTDN0K1lBZ1psWnM4MWNKUT09JkZ0WHdMY1lEcVJVQTMrck9XNEppZWdQaGhsRytnelJESDhtMmtsK1VjRE5tUHhGdGtyTW5LcThTMzlXRTliMXdNUEhHcEhzdXJlRjhJVlZESUI1TCY4Tkg3dUFoc2pkdWVaTGl3NkRwd2lBPT0='),('authentication_ok'),('l8ZnBvotmoGsueFXTOW1WZEfW7VxJ+OhWoZZLl0YR4Ltk5Y5KyX0QSo+kygKhnu3dfjsldfB4HRr5C2koU5yWg=='),('��-��'),('{
  "type" : "POWERAUTH",
  "authDataMethod" : "POST",
  "authDataUriId" : "/hello/biohacker"
}'),('POSSESSION_BIOMETRY'),('4.0'),('2025-08-28 13:51:08.74+02'),('TRUE'::boolean),('4'::int4),('3'::int8)) was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9  Call getNextException to see other errors in the batch.
2025-08-28T13:51:08.782+02:00 ERROR 5746 --- [powerauth-java-server] [nio-8080-exec-3] [4bec8fa5218c990ad990426df7672e8b-07c845616422304a] o.h.engine.jdbc.spi.SqlExceptionHelper   : ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9
2025-08-28T13:51:08.793+02:00 ERROR 5746 --- [powerauth-java-server] [nio-8080-exec-3] [                                                 ] o.s.b.w.servlet.support.ErrorPageFilter  : Forwarding to error page from request [/rest/v4/auth/verify] due to exception [could not execute batch [Batch entry 0 insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (('06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0'),('2'::int8),('xL7oBxcFPIH0IoAFoTm1gQmgmam1UbRZAhFQj9NS6zw='),('3'::int4),('[ ]'),('UE9TVCZMMmhsYkd4dkwySnBiMmhoWTJ0bGNnPT0mT0JrdGtoTDN0K1lBZ1psWnM4MWNKUT09JkZ0WHdMY1lEcVJVQTMrck9XNEppZWdQaGhsRytnelJESDhtMmtsK1VjRE5tUHhGdGtyTW5LcThTMzlXRTliMXdNUEhHcEhzdXJlRjhJVlZESUI1TCY4Tkg3dUFoc2pkdWVaTGl3NkRwd2lBPT0='),('authentication_ok'),('l8ZnBvotmoGsueFXTOW1WZEfW7VxJ+OhWoZZLl0YR4Ltk5Y5KyX0QSo+kygKhnu3dfjsldfB4HRr5C2koU5yWg=='),('��-��'),('{
  "type" : "POWERAUTH",
  "authDataMethod" : "POST",
  "authDataUriId" : "/hello/biohacker"
}'),('POSSESSION_BIOMETRY'),('4.0'),('2025-08-28 13:51:08.74+02'),('TRUE'::boolean),('4'::int4),('3'::int8)) was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9  Call getNextException to see other errors in the batch.] [insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]; SQL [insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]]

org.springframework.dao.DataIntegrityViolationException: could not execute batch [Batch entry 0 insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (('06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0'),('2'::int8),('xL7oBxcFPIH0IoAFoTm1gQmgmam1UbRZAhFQj9NS6zw='),('3'::int4),('[ ]'),('UE9TVCZMMmhsYkd4dkwySnBiMmhoWTJ0bGNnPT0mT0JrdGtoTDN0K1lBZ1psWnM4MWNKUT09JkZ0WHdMY1lEcVJVQTMrck9XNEppZWdQaGhsRytnelJESDhtMmtsK1VjRE5tUHhGdGtyTW5LcThTMzlXRTliMXdNUEhHcEhzdXJlRjhJVlZESUI1TCY4Tkg3dUFoc2pkdWVaTGl3NkRwd2lBPT0='),('authentication_ok'),('l8ZnBvotmoGsueFXTOW1WZEfW7VxJ+OhWoZZLl0YR4Ltk5Y5KyX0QSo+kygKhnu3dfjsldfB4HRr5C2koU5yWg=='),('��-��'),('{
  "type" : "POWERAUTH",
  "authDataMethod" : "POST",
  "authDataUriId" : "/hello/biohacker"
}'),('POSSESSION_BIOMETRY'),('4.0'),('2025-08-28 13:51:08.74+02'),('TRUE'::boolean),('4'::int4),('3'::int8)) was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9  Call getNextException to see other errors in the batch.] [insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]; SQL [insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:297) ~[spring-orm-6.2.9.jar:6.2.9]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.convertHibernateAccessException(HibernateJpaDialect.java:256) ~[spring-orm-6.2.9.jar:6.2.9]
	at org.springframework.orm.jpa.vendor.HibernateJpaDialect.translateExceptionIfPossible(HibernateJpaDialect.java:241) ~[spring-orm-6.2.9.jar:6.2.9]
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:566) ~[spring-orm-6.2.9.jar:6.2.9]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.processCommit(AbstractPlatformTransactionManager.java:795) ~[spring-tx-6.2.9.jar:6.2.9]
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.commit(AbstractPlatformTransactionManager.java:758) ~[spring-tx-6.2.9.jar:6.2.9]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.commitTransactionAfterReturning(TransactionAspectSupport.java:698) ~[spring-tx-6.2.9.jar:6.2.9]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:416) ~[spring-tx-6.2.9.jar:6.2.9]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.9.jar:6.2.9]
	at com.wultra.security.powerauth.app.server.service.behavior.tasks.v4.OnlineAuthenticationServiceBehavior$$SpringCGLIB$$0.verifyAuthentication(<generated>) ~[classes/:2.0.0-SNAPSHOT]
	at com.wultra.security.powerauth.app.server.controller.api.v4.AuthenticationController.verifyAuthentication(AuthenticationController.java:78) ~[classes/:2.0.0-SNAPSHOT]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:360) ~[spring-aop-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.2.9.jar:6.2.9]
	at org.springframework.validation.beanvalidation.MethodValidationInterceptor.invoke(MethodValidationInterceptor.java:174) ~[spring-context-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.2.9.jar:6.2.9]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:728) ~[spring-aop-6.2.9.jar:6.2.9]
	at com.wultra.security.powerauth.app.server.controller.api.v4.AuthenticationController$$SpringCGLIB$$0.verifyAuthentication(<generated>) ~[classes/:2.0.0-SNAPSHOT]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:258) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:191) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:991) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:896) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:590) ~[servlet-api.jar:6.0]
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658) ~[servlet-api.jar:6.0]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51) ~[tomcat-websocket.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.servlet.resource.ResourceUrlEncodingFilter.doFilter(ResourceUrlEncodingFilter.java:66) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:108) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:108) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.web.FilterChainProxy.lambda$doFilterInternal$3(FilterChainProxy.java:231) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$FilterObservation$SimpleFilterObservation.lambda$wrap$1(ObservationFilterChainDecorator.java:479) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$AroundFilterObservation$SimpleAroundFilterObservation.lambda$wrap$1(ObservationFilterChainDecorator.java:340) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator.lambda$wrapSecured$0(ObservationFilterChainDecorator.java:82) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:128) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:125) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:100) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:179) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:107) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:93) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:82) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.context.SecurityContextHolderFilter.doFilter(SecurityContextHolderFilter.java:69) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:62) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:227) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.session.DisableEncodeUrlFilter.doFilterInternal(DisableEncodeUrlFilter.java:42) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.wrapFilter(ObservationFilterChainDecorator.java:240) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$AroundFilterObservation$SimpleAroundFilterObservation.lambda$wrap$0(ObservationFilterChainDecorator.java:323) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$ObservationFilter.doFilter(ObservationFilterChainDecorator.java:224) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.ObservationFilterChainDecorator$VirtualFilterChain.doFilter(ObservationFilterChainDecorator.java:137) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:233) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:191) ~[spring-security-web-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:113) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.ServletRequestPathFilter.doFilter(ServletRequestPathFilter.java:52) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:113) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:74) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.config.annotation.web.configuration.WebSecurityConfiguration$CompositeFilterChainProxy.doFilter(WebSecurityConfiguration.java:319) ~[spring-security-config-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:113) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.servlet.handler.HandlerMappingIntrospector.lambda$createCacheFilter$4(HandlerMappingIntrospector.java:267) ~[spring-webmvc-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.CompositeFilter$VirtualFilterChain.doFilter(CompositeFilter.java:113) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.CompositeFilter.doFilter(CompositeFilter.java:74) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.security.config.annotation.web.configuration.WebMvcSecurityConfiguration$CompositeFilterChainProxy.doFilter(WebMvcSecurityConfiguration.java:240) ~[spring-security-config-6.5.2.jar:6.5.2]
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:362) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:278) ~[spring-web-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.filter.ServerHttpObservationFilter.doFilterInternal(ServerHttpObservationFilter.java:110) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:124) ~[spring-boot-3.5.4.jar:3.5.4]
	at org.springframework.boot.web.servlet.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:99) ~[spring-boot-3.5.4.jar:3.5.4]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:117) ~[spring-boot-3.5.4.jar:3.5.4]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201) ~[spring-web-6.2.9.jar:6.2.9]
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116) ~[spring-web-6.2.9.jar:6.2.9]
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90) ~[catalina.jar:10.1.43]
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:116) ~[catalina.jar:10.1.43]
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93) ~[catalina.jar:10.1.43]
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:666) ~[catalina.jar:10.1.43]
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) ~[catalina.jar:10.1.43]
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344) ~[catalina.jar:10.1.43]
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:398) ~[tomcat-coyote.jar:10.1.43]
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63) ~[tomcat-coyote.jar:10.1.43]
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:903) ~[tomcat-coyote.jar:10.1.43]
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1769) ~[tomcat-coyote.jar:10.1.43]
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52) ~[tomcat-coyote.jar:10.1.43]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1189) ~[tomcat-util.jar:10.1.43]
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:658) ~[tomcat-util.jar:10.1.43]
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63) ~[tomcat-util.jar:10.1.43]
	at java.base/java.lang.Thread.run(Thread.java:833) ~[na:na]
Caused by: org.hibernate.exception.DataException: could not execute batch [Batch entry 0 insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (('06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0'),('2'::int8),('xL7oBxcFPIH0IoAFoTm1gQmgmam1UbRZAhFQj9NS6zw='),('3'::int4),('[ ]'),('UE9TVCZMMmhsYkd4dkwySnBiMmhoWTJ0bGNnPT0mT0JrdGtoTDN0K1lBZ1psWnM4MWNKUT09JkZ0WHdMY1lEcVJVQTMrck9XNEppZWdQaGhsRytnelJESDhtMmtsK1VjRE5tUHhGdGtyTW5LcThTMzlXRTliMXdNUEhHcEhzdXJlRjhJVlZESUI1TCY4Tkg3dUFoc2pkdWVaTGl3NkRwd2lBPT0='),('authentication_ok'),('l8ZnBvotmoGsueFXTOW1WZEfW7VxJ+OhWoZZLl0YR4Ltk5Y5KyX0QSo+kygKhnu3dfjsldfB4HRr5C2koU5yWg=='),('��-��'),('{
  "type" : "POWERAUTH",
  "authDataMethod" : "POST",
  "authDataUriId" : "/hello/biohacker"
}'),('POSSESSION_BIOMETRY'),('4.0'),('2025-08-28 13:51:08.74+02'),('TRUE'::boolean),('4'::int4),('3'::int8)) was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9  Call getNextException to see other errors in the batch.] [insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)]
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:103) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:58) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:108) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:293) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.mutation.internal.PreparedStatementGroupSingleTable.forEachStatement(PreparedStatementGroupSingleTable.java:67) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.performExecution(BatchImpl.java:264) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.execute(BatchImpl.java:242) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.executeBatch(JdbcCoordinatorImpl.java:188) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:674) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:511) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:414) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:41) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.event.service.internal.EventListenerGroupImpl.fireEventOnEachListener(EventListenerGroupImpl.java:127) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1429) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:491) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:2354) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:1978) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:439) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:169) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:267) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:101) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	at org.springframework.orm.jpa.JpaTransactionManager.doCommit(JpaTransactionManager.java:562) ~[spring-orm-6.2.9.jar:6.2.9]
	... 151 common frames omitted
Caused by: java.sql.BatchUpdateException: Batch entry 0 insert into pa_signature_audit (activation_id,activation_counter,activation_ctr_data,activation_status,additional_info,data_base64,note,signature,signature_data_body,signature_metadata,signature_type,signature_version,timestamp_created,valid,version,id) values (('06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0'),('2'::int8),('xL7oBxcFPIH0IoAFoTm1gQmgmam1UbRZAhFQj9NS6zw='),('3'::int4),('[ ]'),('UE9TVCZMMmhsYkd4dkwySnBiMmhoWTJ0bGNnPT0mT0JrdGtoTDN0K1lBZ1psWnM4MWNKUT09JkZ0WHdMY1lEcVJVQTMrck9XNEppZWdQaGhsRytnelJESDhtMmtsK1VjRE5tUHhGdGtyTW5LcThTMzlXRTliMXdNUEhHcEhzdXJlRjhJVlZESUI1TCY4Tkg3dUFoc2pkdWVaTGl3NkRwd2lBPT0='),('authentication_ok'),('l8ZnBvotmoGsueFXTOW1WZEfW7VxJ+OhWoZZLl0YR4Ltk5Y5KyX0QSo+kygKhnu3dfjsldfB4HRr5C2koU5yWg=='),('��-��'),('{
  "type" : "POWERAUTH",
  "authDataMethod" : "POST",
  "authDataUriId" : "/hello/biohacker"
}'),('POSSESSION_BIOMETRY'),('4.0'),('2025-08-28 13:51:08.74+02'),('TRUE'::boolean),('4'::int4),('3'::int8)) was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9  Call getNextException to see other errors in the batch.
	at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:165) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2422) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:580) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.jdbc.PgStatement.internalExecuteBatch(PgStatement.java:886) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:910) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1778) ~[postgresql-42.7.7.jar:42.7.7]
	at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:128) ~[HikariCP-6.3.1.jar:na]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java) ~[HikariCP-6.3.1.jar:na]
	at org.hibernate.engine.jdbc.batch.internal.BatchImpl.lambda$performExecution$2(BatchImpl.java:279) ~[hibernate-core-6.6.22.Final.jar:6.6.22.Final]
	... 169 common frames omitted
Caused by: org.postgresql.util.PSQLException: ERROR: invalid byte sequence for encoding "UTF8": 0x00
  Where: unnamed portal parameter $9
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2734) ~[postgresql-42.7.7.jar:42.7.7]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2421) ~[postgresql-42.7.7.jar:42.7.7]
	... 176 common frames omitted

2025-08-28T13:51:09.068+02:00 DEBUG 5746 --- [powerauth-java-server] [   scheduling-1] [bf2e7c1a560bab87559ee79f07482b32-14dc3b836acc5334] c.w.c.a.b.database.DatabaseAuditWriter   : Scheduled audit log flush called
2025-08-28T13:51:09.071+02:00 DEBUG 5746 --- [powerauth-java-server] [   scheduling-1] [bf2e7c1a560bab87559ee79f07482b32-14dc3b836acc5334] c.w.c.a.b.database.DatabaseAuditWriter   : Audit log batch insert succeeded, audit record count: 1, audit param is disabled
```

--- a/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/annotation/Base64Encoded.java
+++ b/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/annotation/Base64Encoded.java
@@ -1,0 +1,50 @@
+/*
+ * PowerAuth Server and related software components
+ * Copyright (C) 2025 Wultra s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.wultra.security.powerauth.client.model.annotation;
+
+import com.wultra.security.powerauth.client.validation.Base64Validator;
+import jakarta.validation.Constraint;
+import jakarta.validation.Payload;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+/**
+ * Base-64 encoded string annotation for model classes validation.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Target(FIELD)
+@Retention(RUNTIME)
+@Constraint(validatedBy = Base64Validator.class)
+@Documented
+public @interface Base64Encoded {
+
+    String message() default "Must be a valid base64 string";
+
+    Class<?>[] groups() default {};
+
+    Class<? extends Payload>[] payload() default {};
+
+}

--- a/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v3/VerifyECDSASignatureRequest.java
+++ b/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v3/VerifyECDSASignatureRequest.java
@@ -19,6 +19,7 @@
 
 package com.wultra.security.powerauth.client.model.request.v3;
 
+import com.wultra.security.powerauth.client.model.annotation.Base64Encoded;
 import com.wultra.security.powerauth.client.model.enumeration.v3.ECDSASignatureFormat;
 import io.swagger.v3.oas.annotations.media.Schema;
 import jakarta.validation.constraints.NotBlank;
@@ -44,6 +45,7 @@ public class VerifyECDSASignatureRequest {
 
     @Schema(description = "Signature")
     @NotBlank(message = "Signature must not be empty when verifying signature")
+    @Base64Encoded(message = "Signature must be a valid Base-64 encoded string")
     @ToString.Exclude
     private String signature;
 

--- a/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v4/VerifyAsymmetricSignatureRequest.java
+++ b/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v4/VerifyAsymmetricSignatureRequest.java
@@ -19,6 +19,7 @@
 
 package com.wultra.security.powerauth.client.model.request.v4;
 
+import com.wultra.security.powerauth.client.model.annotation.Base64Encoded;
 import com.wultra.security.powerauth.client.model.enumeration.v4.AsymmetricSignatureFormat;
 import com.wultra.security.powerauth.client.model.enumeration.v4.AsymmetricSignatureType;
 import io.swagger.v3.oas.annotations.media.Schema;
@@ -45,6 +46,7 @@ public class VerifyAsymmetricSignatureRequest {
 
     @Schema(description = "Signature")
     @NotBlank(message = "Signature must not be empty when verifying signature")
+    @Base64Encoded(message = "Signature must be a valid Base-64 encoded string")
     @ToString.Exclude
     private String signature;
 

--- a/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v4/VerifyAuthenticationRequest.java
+++ b/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/v4/VerifyAuthenticationRequest.java
@@ -19,6 +19,7 @@
 
 package com.wultra.security.powerauth.client.model.request.v4;
 
+import com.wultra.security.powerauth.client.model.annotation.Base64Encoded;
 import com.wultra.security.powerauth.client.model.enumeration.ActivationStatus;
 import com.wultra.security.powerauth.client.model.enumeration.v4.AuthenticationCodeType;
 import io.swagger.v3.oas.annotations.media.Schema;
@@ -52,6 +53,7 @@ public class VerifyAuthenticationRequest {
 
     @Schema(description = "Authentication code")
     @NotBlank(message = "Authentication code must not be empty when verifying authentication")
+    @Base64Encoded(message = "Authentication code must be a valid Base-64 encoded string")
     @ToString.Exclude
     private String authenticationCode;
 

--- a/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/validation/Base64Validator.java
+++ b/powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/validation/Base64Validator.java
@@ -1,0 +1,47 @@
+/*
+ * PowerAuth Server and related software components
+ * Copyright (C) 2025 Wultra s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.wultra.security.powerauth.client.validation;
+
+import com.wultra.security.powerauth.client.model.annotation.Base64Encoded;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+
+import java.util.Base64;
+
+/**
+ * Validator for Base-64 encoded strings.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+public class Base64Validator implements ConstraintValidator<Base64Encoded, String> {
+
+    @Override
+    public boolean isValid(String value, ConstraintValidatorContext context) {
+        if (value == null || value.isBlank()) {
+            return true;
+        }
+        try {
+            Base64.getDecoder().decode(value);
+            return true;
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+    }
+}

--- a/powerauth-java-server/src/test/java/com/wultra/security/powerauth/app/server/controller/validation/VerifyAuthenticationRequestValidationTest.java
+++ b/powerauth-java-server/src/test/java/com/wultra/security/powerauth/app/server/controller/validation/VerifyAuthenticationRequestValidationTest.java
@@ -1,0 +1,103 @@
+/*
+ * PowerAuth Server and related software components
+ * Copyright (C) 2025 Wultra s.r.o.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.wultra.security.powerauth.app.server.controller.validation;
+
+import com.wultra.security.powerauth.client.model.enumeration.v4.AuthenticationCodeType;
+import com.wultra.security.powerauth.client.model.request.v4.VerifyAuthenticationRequest;
+import jakarta.validation.ConstraintViolation;
+import jakarta.validation.Validation;
+import jakarta.validation.Validator;
+import jakarta.validation.ValidatorFactory;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+import java.util.Set;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test for authentication code validation.
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+class VerifyAuthenticationRequestValidationTest {
+
+    private static ValidatorFactory factory;
+    private static Validator validator;
+
+    @BeforeAll
+    static void setupValidator() {
+        factory = Validation.buildDefaultValidatorFactory();
+        validator = factory.getValidator();
+    }
+
+    @AfterAll
+    static void tearDownValidator() {
+        factory.close();
+    }
+
+    private VerifyAuthenticationRequest buildBaseRequest(String authenticationCode) {
+        final VerifyAuthenticationRequest req = new VerifyAuthenticationRequest();
+        req.setActivationId("06c4d2d4-4bff-4ea7-b58c-5cfe8b7f0bb0");
+        req.setApplicationKey("qjsBj6II/syLPs65BbNRyg==");
+        req.setData("POST&L3BhL2xvZ2lu&vkueT796IGqdXlfVIJrB9A==&eyJQbGF0Zm9ybSI6ICJBIiwiTGFuZ3VhZ2VDb2RlIjogIkVORyJ9");
+        req.setAuthenticationCode(authenticationCode);
+        req.setAuthenticationCodeType(AuthenticationCodeType.POSSESSION_BIOMETRY);
+        req.setAuthenticationVersion("4.0");
+        return req;
+    }
+
+    @Test
+    void validBase64AuthenticationCode_passesValidation() {
+        final VerifyAuthenticationRequest req = buildBaseRequest("f0LKq3AVXO4unDIKhclgLfaxeMz4YBFJokAx8ZB6t2/ZfGzIQzdLHClynGqCNQ9O/0OUzOwWYEK/cS/CCJFqX4WkR8kRXpSFNz8BqmfkJwkhwS5WE1OTq+1P02GC6C0R");
+        final Set<ConstraintViolation<VerifyAuthenticationRequest>> violations = validator.validate(req);
+        assertTrue(violations.isEmpty());
+    }
+
+    @Test
+    void invalidBase64AuthenticationCode_failsValidation() {
+        final VerifyAuthenticationRequest req = buildBaseRequest("XXX~");
+        final Set<ConstraintViolation<VerifyAuthenticationRequest>> violations = validator.validate(req);
+        assertFalse(violations.isEmpty(), "Expected violations for an invalid Base64 authentication code");
+        assertTrue(violations.stream().anyMatch(v -> "authenticationCode".equals(v.getPropertyPath().toString())));
+    }
+
+    @Test
+    void binaryLikeStringWithNul_failsValidation() {
+        final String binary = "abc\u0000def";
+        final VerifyAuthenticationRequest req = buildBaseRequest(binary);
+        final Set<ConstraintViolation<VerifyAuthenticationRequest>> violations = validator.validate(req);
+        assertFalse(violations.isEmpty());
+        assertTrue(violations.stream().anyMatch(v -> "authenticationCode".equals(v.getPropertyPath().toString())));
+    }
+
+    @Test
+    void blankAuthenticationCode_triggersNotBlank() {
+        final VerifyAuthenticationRequest req = buildBaseRequest("   ");
+        final Set<ConstraintViolation<VerifyAuthenticationRequest>> violations = validator.validate(req);
+        assertFalse(violations.isEmpty());
+        assertTrue(violations.stream().anyMatch(v ->
+                        "authenticationCode".equals(v.getPropertyPath().toString()) && v.getMessage().equals("Authentication code must not be empty when verifying authentication"))
+        );
+    }
+
+}

